mkosi/__init__.py

```diff
@@ -531,15 +531,16 @@ def gpt_root_native(arch: Optional[str], usr_only: bool = False) -> GPTRootTypeT
             die(f"Unknown architecture {arch}.")
 
 
-def roothash_suffix(usr_only: bool = False) -> str:
-    if usr_only:
-        return ".usrhash"
+def root_or_usr(args: Union[MkosiArgs, argparse.Namespace]) -> str:
+    return ".usr" if args.usr_only else ".root"
+
 
-    return ".roothash"
+def roothash_suffix(args: Union[MkosiArgs, argparse.Namespace]) -> str:
+    return root_or_usr(args) + root_or_usr(args) + "hash"
```

Comment:
It is necessary. This is messy because of a mistake in systemd: the 'root' in roothash never referred to the partition mount point, it simply means that the content is the hash of the root element of the verity merkle tree. Unfortunately for /usr it was called usrhash - that was a mistake that we have to live with.

As the commit message says, systemd and various tools auto discover verity artifacts from an image name: foo.raw will search for foo.roothash, foo.roothash.p7s, foo.verity (same for usr). So the prefix has to be exactly the same. You can't have foo.root.raw and foo.roothash, it wouldn't work.

Then we need to be able to have both usr and root artifacts in the same directory otherwise they'll overwrite each other and get messy.

Hence these naming choices.

Reply:
> As the commit message says, systemd and various tools auto discover verity artifacts from an image name: foo.raw will search for foo.roothash, foo.roothash.p7s, foo.verity (same for usr). So the prefix has to be exactly the same. You can't have foo.root.raw and foo.roothash, it wouldn't work.

Yes, that much was clear from the `systemd.exec` man page, but...

> Then we need to be able to have both usr and root artifacts in the same directory otherwise they'll overwrite each other and get messy.

...I think that's where my confusion came from, because I thought for one input we would have either a root or a usr artifact, but not both (and for different inputs I would usually choose or already have separate output directories).

Reply:
Removing the .usr. .root. suffixes would be a compatibility break no? Also we don't enforce a different output directory when one line in the config has changed, so we can't assume it will.